pkg/scalers/etcd_scaler.go

```diff
@@ -0,0 +1,306 @@
+			keyGiven := config.AuthParams["key"] != ""
+			if certGiven && !keyGiven {
+				return errors.New("key must be provided with cert")
+			}
+			if keyGiven && !certGiven {
+				return errors.New("cert must be provided with key")
+			}
+			meta.ca = config.AuthParams["ca"]
+			meta.cert = config.AuthParams["cert"]
+			meta.key = config.AuthParams["key"]
+			if value, found := config.AuthParams["keyPassword"]; found {
+				meta.keyPassword = value
+			} else {
+				meta.keyPassword = ""
+			}
+			meta.enableTLS = true
+		} else if val != etcdTLSDisable {
+			return fmt.Errorf("err incorrect value for TLS given: %s", val)
+		}
+	}
+
+	return nil
+}
+
+func parseEtcdMetadata(config *ScalerConfig) (*etcdMetadata, error) {
+	meta := &etcdMetadata{}
+	var err error
+	meta.etcdEndpoints = strings.Split(config.TriggerMetadata[etcdEndpoints], ",")
+	if len(meta.etcdEndpoints) == 0 || meta.etcdEndpoints[0] == "" {
+		return nil, fmt.Errorf("etcdEndpoints required")
+	}
+
+	meta.etcdKey = config.TriggerMetadata[etcdKey]
+	if len(meta.etcdKey) == 0 {
+		return nil, fmt.Errorf("etcdKey required")
+	}
+
+	meta.value, err = strconv.ParseFloat(config.TriggerMetadata[value], 64)
+	if err != nil || meta.value <= 0 {
+		return nil, fmt.Errorf("value must be a float greater than 0")
+	}
+
+	meta.activationValue = 0
+	if val, ok := config.TriggerMetadata[activationValue]; ok {
+		meta.activationValue, err = strconv.ParseFloat(val, 64)
+		if err != nil {
+			return nil, fmt.Errorf("activationValue must be a float")
+		}
+	}
+
+	meta.watchProgressNotifyInterval = 600
+	if val, ok := config.TriggerMetadata[watchProgressNotifyInterval]; ok {
+		meta.watchProgressNotifyInterval, err = strconv.Atoi(val)
+		if err != nil || meta.watchProgressNotifyInterval <= 0 {
+			return nil, fmt.Errorf("watchProgressNotifyInterval must be a int greater than 0")
+		}
+	}
+
+	if err = parseEtcdAuthParams(config, meta); err != nil {
+		return meta, err
+	}
+
+	meta.scalerIndex = config.ScalerIndex
+	return meta, nil
+}
+
+func getEtcdClients(metadata *etcdMetadata) (*clientv3.Client, error) {
+	var tlsConfig *tls.Config
+	var err error
+	if metadata.enableTLS {
+		tlsConfig, err = kedautil.NewTLSConfigWithPassword(metadata.cert, metadata.key, metadata.keyPassword, metadata.ca)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   metadata.etcdEndpoints,
+		DialTimeout: 5 * time.Second,
+		TLS:         tlsConfig,
+	})
+	fmt.Println(metadata.etcdEndpoints)
+	if err != nil {
+		return nil, fmt.Errorf("error connecting to etcd server: %s", err)
+	}
+
+	return cli, nil
+}
+
+// IsActive determines if we need to scale from zero
+func (s *etcdScaler) IsActive(ctx context.Context) (bool, error) {
+	v, err := s.getMetricValue(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	return v > s.metadata.activationValue, nil
+}
+
+// Close closes the etcd client
+func (s *etcdScaler) Close(context.Context) error {
+	return s.client.Close()
```

Comment:
I'm not 100% sure if we shouldn't check if `client != nil` before calling `Close`. I'd say that there are some cases where we can panic here calling to close without initialising the object before

Reply:
Agree！We can not rule out the possibility of a null pointer.